### PR TITLE
Added support for importing Parquet complex types

### LINKED ISSUE #11: Complex types in Parquet storage format
Exasol does not support complex types such as `Map`, `Struct` or `Array`. 

Therefore, it is important how to support complex types in Parquet or Avro formats.

One possible option is to converting them into json string and to store them as a `VARCHAR` in Exasol table.

### PR DESCRIPTION
**Pull request by morazow**
Fixes #11.

### CODE REVIEW

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ValueHolder.scala [hunk 1]
```diff
@@ -1,12 +1,14 @@
 package com.exasol.cloudetl.parquet.converter
 
+import scala.collection.mutable.ArrayBuffer
+
 /**
  * An interface for the storing converted Parquet values.
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/parquet/converter/ValueHolder.scala:6]:
```suggestion
 * An interface for storing the converted Parquet values.
```

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala [hunk 2]
```diff
@@ -134,3 +140,219 @@ final case class ParquetDateConverter(index: Int, holder: ValueHolder)
...
+  private[this] def createFieldConverters(): Array[Converter] = {
+    val converters = Array.ofDim[Converter](size)
+    for { i <- 0 until size } {
+      converters(i) = ConverterFactory(groupType.getType(i), i, dataHolder)
+    }
+    converters
+  }
+}
+
+final case class StructConverter(groupType: GroupType, index: Int, parentDataHolder: ValueHolder)
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:344]: add doc to all converters. I don't get the different purposes

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: src/test/scala/com/exasol/cloudetl/parquet/converter/ParquetConverterTest.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+         |  }
+         |}
+         |""".stripMargin
+    )
+    Seq(-1, 1, 2, 5).foreach {
+      case index =>
+        val thrown = intercept[IllegalArgumentException] {
+          ArrayGroupConverter(schema.getType(0), -1, EmptyValueHolder).getConverter(index)
+        }
+        assert(thrown.getMessage().contains(s"Illegal index '$index' to array converter"))
```
[the remark this answers is not in the document]
Comment (morazow) [reply src/test/scala/com/exasol/cloudetl/parquet/converter/ParquetConverterTest.scala:24]: Yes, I can use it. I am going to change all in one issue #113.

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala [hunk 4]
```diff
@@ -134,3 +140,219 @@ final case class ParquetDateConverter(index: Int, holder: ValueHolder)
...
+        keysConverter
+      } else {
+        valuesConverter
+      }
+    override def start(): Unit = {}
+    override def end(): Unit = {}
+  }
+}
+
+@SuppressWarnings(Array("org.wartremover.contrib.warts.UnsafeInheritance"))
```
[the remark this answers is not in the document]
Comment (morazow) [reply src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:324]: Fixed properly, removed the suppression.

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala [hunk 5]
```diff
@@ -134,3 +140,219 @@ final case class ParquetDateConverter(index: Int, holder: ValueHolder)
     holder.put(index, date)
   }
 }
+
+final case class RepeatedGroupConverter(
```
[the remark this answers is not in the document]
Comment (morazow) [reply src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:144]: Fixed.

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala (hunk 2 above)
Comment (morazow) [reply src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:344]: Fixed.

File: doc/changes/changes_0.9.1.md [hunk 6]
```diff
@@ -27,7 +28,7 @@
 ### Test Dependency Updates
 
 * Updates `org.scalatest:scalatest` from `3.2.2` to `3.2.3`.
-* Updated `org.mockito:mockito-core` from `3.5.13` to `3.6.0`.
+* Updated `org.mockito:mockito-core` from `3.5.13` to `3.6.28`.
```
[the remark this answers is not in the document]
Comment (morazow) [reply doc/changes/changes_0.9.1.md:31]: Fixed.

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala [hunk 7]
```diff
@@ -23,11 +23,31 @@ import org.apache.parquet.schema.Type
 /**
  * An interface for the Parquet data type converters.
  *
+ * The Parquet reader calls the [[ParquetRootConverter]] for the top
+ * level Parquet schema. The root converter then generates subsequent
+ * converters using [[ConverterFactory]] for each type.
+ *
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:29]: :+1:

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala [hunk 8]
```diff
@@ -60,6 +95,23 @@ final case class ParquetStringConverter(index: Int, holder: ValueHolder)
     holder.put(index, decodedDictionary(dictionaryId))
 }
 
+/**
+ * A converter for {@code DECIMAL} annotated Parquet types.
+ *
+ * The decimal annotation can be used for the following Parquet types:
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:101]: In scala the formating is kept without `<p>`?

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala [hunk 9]
```diff
@@ -60,6 +95,23 @@ final case class ParquetStringConverter(index: Int, holder: ValueHolder)
...
+ * {@code INT32}, {@code INT64}, {@code FIXED_LEN_BYTE_ARRAY} and {@code
+ * BINARY}.
+ *
+ * The following schema fits this converter:
+ * {{{
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:106]: That's also scala specific?

**Review by jakobbraun — APPROVED** [review]
approved with two questions

File: src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala (hunk 9 above)
Comment (morazow) [reply src/main/scala/com/exasol/cloudetl/parquet/converter/ParquetConverter.scala:106]: Yes, exactly. It has it is own renderer. `{{{` can be used for larger code blocks.

An example api doc page:
Screenshot from 2020-11-30 11-20-12

**Pull request merged**